In [1]:
import numpy as np
import pandas as pd
import joblib
import pickle as pkl
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

### chargement de données

In [2]:
#X_train, X_test, y_train, y_test = joblib.load("data_split.pkl")

In [5]:
y_train = joblib.load("y_train.joblib")

In [6]:
X_train = joblib.load("X_train.joblib")

In [4]:
X_test = joblib.load("X_test.joblib")

In [5]:
y_test = joblib.load("y_test.joblib")

### gridsearch for SVC

In [6]:
# 解説5：グリッドサーチを設定する---------------------------------------------------
#clf=SVC(class_weight='balanced', random_state=42)
clf=SVC(random_state=42)

param_range1=[0.001,0.01] #変化させるパラメータに設定する値たち
param_range2=[10,100,1000]
hyper_param_grid=[{'C':param_range2,
             'kernel':['rbf'], 
             'gamma':param_range1}] #Cとカーネルとgammaを変化させて最適化させる

clf_grid_search = GridSearchCV(
    estimator = clf, 
    param_grid = hyper_param_grid, 
    cv=3, 
    return_train_score=True,
    verbose=2)

clf_grid_search.fit(X_train, y_train)

Fitting 3 folds for each of 6 candidates, totalling 18 fits
[CV] END ......................C=10, gamma=0.001, kernel=rbf; total time=25.9min
[CV] END ......................C=10, gamma=0.001, kernel=rbf; total time=27.4min
[CV] END ......................C=10, gamma=0.001, kernel=rbf; total time=28.7min
[CV] END .......................C=10, gamma=0.01, kernel=rbf; total time=87.8min
[CV] END .......................C=10, gamma=0.01, kernel=rbf; total time=87.0min
[CV] END .......................C=10, gamma=0.01, kernel=rbf; total time=90.7min
[CV] END .....................C=100, gamma=0.001, kernel=rbf; total time=48.8min
[CV] END .....................C=100, gamma=0.001, kernel=rbf; total time=48.0min
[CV] END .....................C=100, gamma=0.001, kernel=rbf; total time=48.6min
[CV] END ......................C=100, gamma=0.01, kernel=rbf; total time=88.2min
[CV] END ......................C=100, gamma=0.01, kernel=rbf; total time=99.7min
[CV] END ......................C=100, gamma=0.01,

GridSearchCV(cv=3, estimator=SVC(random_state=42),
             param_grid=[{'C': [10, 100, 1000], 'gamma': [0.001, 0.01],
                          'kernel': ['rbf']}],
             return_train_score=True, verbose=2)

In [12]:
results=pd.DataFrame(clf_grid_search.cv_results_)
results.T

0  \
mean_fit_time                                       935.323572   
std_fit_time                                         54.219685   
mean_score_time                                     704.999566   
std_score_time                                       30.548035   
param_C                                                     10   
param_gamma                                              0.001   
param_kernel                                               rbf   
params              {'C': 10, 'gamma': 0.001, 'kernel': 'rbf'}   
split0_test_score                                     0.883489   
split1_test_score                                     0.879275   
split2_test_score                                     0.884795   
mean_test_score                                        0.88252   
std_test_score                                        0.002356   
rank_test_score                                              4   
split0_train_score                                    0.958718   
split1_train_score                                    0.960221   
split2_train_score                                    0.959021   
mean_train_score                                       0.95932   
std_train_score                                       0.000649   

                                                            1  \
mean_fit_time                                      4193.65592   
std_fit_time                                        59.723849   
mean_score_time                                    1118.76803   
std_score_time                                        38.3213   
param_C                                                    10   
param_gamma                                              0.01   
param_kernel                                              rbf   
params              {'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}   
split0_test_score                                    0.883849   
split1_test_score                                    0.880475   
split2_test_score                                    0.885275   
mean_test_score                                        0.8832   
std_test_score                                       0.002013   
rank_test_score                                             3   
split0_train_score                                        1.0   
split1_train_score                                        1.0   
split2_train_score                                        1.0   
mean_train_score                                          1.0   
std_train_score                                           0.0   

                                                              2  \
mean_fit_time                                       2312.319777   
std_fit_time                                          25.393948   
mean_score_time                                      596.532618   
std_score_time                                        36.516442   
param_C                                                     100   
param_gamma                                               0.001   
param_kernel                                                rbf   
params              {'C': 100, 'gamma': 0.001, 'kernel': 'rbf'}   
split0_test_score                                      0.865011   
split1_test_score                                      0.866075   
split2_test_score                                      0.865715   
mean_test_score                                          0.8656   
std_test_score                                         0.000442   
rank_test_score                                               5   
split0_train_score                                      0.99904   
split1_train_score                                      0.99886   
split2_train_score                                      0.99844   
mean_train_score                                        0.99878   
std_train_score                                        0.000251   

                                                             3  \
mean_fit_time                   

In [7]:
# to see the best hyperparameters

clf_grid_search.best_params_

{'C': 100, 'gamma': 0.01, 'kernel': 'rbf'}

In [9]:
# to see the best validation score
clf_grid_search.best_score_

0.8833599516286866

In [15]:
y_train_pred = clf_grid_search.predict(X_train)

In [16]:
accuracy_train = accuracy_score(y_train, y_train_pred)
print(f'train/validation accuracy: {accuracy_train}')

train/validation accuracy: 1.0


In [10]:
# to see the best hyperparameters
clf_svc = clf_grid_search.best_estimator_
clf_svc

SVC(C=100, gamma=0.01, random_state=42)

modèle avec best hyperparameters est over fitting

    train score(accuracy): 1.0
    validation score(accuracy) : 0.8833599516286866

donc c'est mieux de prendre le modèle avec test_rank_score=4

    train score(accuracy): 0.95932
    validation score(accuracy) : 0.88252
    train score(accuracy) - validation score(accuracy) = 0.0768

    SVC(C=10, gamma=0.001, kernel=rbf, random_state=42)

### Focus sur le meilleur modele

#### matrix de confusion

In [18]:
#création de matrix de confusion
cm_train = confusion_matrix(y_train, y_train_pred)

In [19]:
cm_train

array([[12500,     0],
       [    0, 12500]], dtype=int64)

### save the best model avec joblib

In [11]:
%%time
# 構築したモデルの保存
# best_estimator_ over fitting
filename = 'clf_svc.pkl'
joblib.dump(clf_svc,filename)

Wall time: 965 ms


['clf_svc.pkl']

le modéle non over fitting

In [3]:
clf_svc_best = SVC(C=10, gamma=0.001, kernel='rbf', random_state=42)

In [7]:
clf_svc_best.fit(X_train, y_train)

SVC(C=10, gamma=0.001, random_state=42)

In [8]:
%%time
# 構築したモデルの保存
filename = 'clf_svc_best.pkl'
joblib.dump(clf_svc_best,filename)

Wall time: 679 ms


['clf_svc_best.pkl']

https://www.salesanalytics.co.jp/datascience/datascience044/